R/join_keys.R

```diff
@@ -0,0 +1,668 @@
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.TealData <- function(...) {
+  x <- rlang::list2(...)
+  x[[1]]$get_join_keys()
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.default <- function(...) {
+  # Constructor using join_keys<-.xxx setter
+  result <- new_join_keys()
+  join_keys(result) <- rlang::list2(...)
+  result
+}
+
+#' @rdname join_keys
+#' @details
+#' The setter assignment `join_keys(obj) <- ...` will merge obj and `...` if obj
+#' is not empty.
+#'
+#' @param join_keys_obj (`join_keys`) empty object to set the new relationship pairs.
+#' @param value (`join_key_set` or list of `join_key_set`) relationship pairs to add
+#' to `join_keys` list.
+#'
+#' @export
+`join_keys<-` <- function(join_keys_obj, value) {
+  checkmate::assert(
+    combine = "or",
+    checkmate::check_class(value, classes = c("join_keys", "list")),
+    checkmate::check_class(value, classes = c("join_key_set")),
+    checkmate::check_list(value, types = "join_key_set")
+  )
+  UseMethod("join_keys<-", join_keys_obj)
+}
+
+#' @rdname join_keys
+#' @export
+#' @examples
+#'
+#' # Using the setter (assignment) ----
+#'
+#' jk <- join_keys()
+#' join_keys(jk) <- join_key("ds1", "ds2", "some_col")
+#' join_keys(jk) <- c(join_keys(jk), join_key("ds3", "ds4", "some_col2"))
+#'
+#' join_keys(jk)[["ds1"]][["ds3"]] <- "some_col3"
+#' jk
+`join_keys<-.join_keys` <- function(join_keys_obj, value) {
+  # Assume assignment of join keys as a merge operation
+  #  Needed to support join_keys(jk)[c("ds1", "ds2")] <- "key"
+  if (checkmate::test_class(value, classes = c("join_keys", "list"))) {
+    return(value)
+  }
+
+  if (inherits(value, "join_key_set")) value <- list(value)
+
+  join_keys_obj <- new_join_keys()
+
+  # check if any join_key_sets share the same datasets but different values
+  for (idx_1 in seq_along(value)) {
+    for (idx_2 in seq_along(value[idx_1])) {
+      assert_compatible_keys(value[[idx_1]], value[[idx_2]])
+    }
+    dataset_1 <- get_dataset_1(value[[idx_1]])
+    dataset_2 <- get_dataset_2(value[[idx_1]])
+    keys <- get_keys(value[[idx_1]])
+
+    join_keys_obj[[dataset_1]][[dataset_2]] <- keys
+  }
+
+  logger::log_trace("join_keys keys are set.")
+
+  join_keys_obj
+}
+
+#' @rdname join_keys
+#' @export
+#' @examples
+#'
+#' # Setter for join_keys within teal_data ----
+#'
+#' td <- teal_data()
+#' join_keys(td)[["ds1"]][["ds2"]] <- "key1"
+#' join_keys(td)[["ds2"]][["ds2"]] <- "key2"
+#' join_keys(td) <- c(join_keys(td), join_keys(join_key("ds3", "ds2", "key3")))
+#' join_keys(td)
+`join_keys<-.teal_data` <- function(join_keys_obj, value) {
+  join_keys(join_keys_obj@join_keys) <- value
+  join_keys_obj
+}
+
+#' @rdname join_keys
+#' @export
+#'
+#' @examples
+#'
+#' c(join_keys(join_key("a", "b", "c")), join_keys(join_key("a", "d2", "c")))
+c.join_keys <- function(...) {
+  x <- rlang::list2(...)
+  checkmate::assert_list(x[-1], types = c("join_keys", "join_key_set"))
+
+  merge_join_keys.join_keys(x[[1]], x[-1])
```

Comment:
This is the only place where merge_join_keys is called. You can easily remove merge_join_keys method and insert `merge_join_keys.join_keys` code here. Other `merge_join_keys` are redundant